src/qonnx/transformation/resize_conv_to_deconv.py

```diff
@@ -0,0 +1,262 @@
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import numpy as np
+import warnings
+from onnx import helper
+
+from qonnx.core.datatype import DataType
+from qonnx.custom_op.general.quant import quant
+from qonnx.transformation.base import Transformation
+from qonnx.util.basic import get_by_name
+
+
+def _weight_convolution(cnv_weights: np.ndarray, scale: int) -> np.ndarray:
+    """Adaptation of the weight convolution algorithm as proposed in Colbert et al. (2021) - `An
+    Energy-Efficient Edge Computing Paradigm for Convolution-Based Image Upsampling`"""
+    ofm_ch = cnv_weights.shape[0]
+    ifm_ch = cnv_weights.shape[1]
+    kh_size = cnv_weights.shape[2]
+    kw_size = cnv_weights.shape[3]
+    assert kh_size == kw_size, "Only square channels supported currently."
+    # NOTE - this is different than the convolution kernels, which are OC x IC x KH x KW
+    #        rather than IC x OC x KH x KW
+    dcnv_weights = np.zeros((ifm_ch, ofm_ch, kh_size + scale - 1, kw_size + scale - 1))
+    for oc in range(ofm_ch):
+        for ic in range(ifm_ch):
+            for i in range(scale):
+                for j in range(scale):
+                    dcnv_weights[ic, oc, i : i + kh_size, j : j + kw_size] += np.rot90(cnv_weights[oc, ic], 2, [0, 1])
+    return dcnv_weights
+
+
+def _auto_pad_to_explicit_padding(autopad_str, idim_h, idim_w, k_h, k_w, stride_h, stride_w, n_dims):
+    pad_total_h = (stride_h - 1) * idim_h - stride_h + k_h
+    pad_total_w = (stride_w - 1) * idim_w - stride_w + k_w
+    pad_half_small_h = int((pad_total_h / 2))
+    pad_half_small_w = int((pad_total_w / 2))
+    pad_half_large_h = pad_total_h - pad_half_small_h
+    pad_half_large_w = pad_total_w - pad_half_small_w
+    if autopad_str == "VALID":
+        return [0 for i in range(2 * n_dims)]
+    elif autopad_str == "SAME_UPPER":
+        return [pad_half_small_h, pad_half_small_w, pad_half_large_h, pad_half_large_w]
+    elif autopad_str == "SAME_LOWER":
+        return [pad_half_large_h, pad_half_large_w, pad_half_small_h, pad_half_small_w]
+    else:
+        raise Exception("Unsupported auto_pad: " + autopad_str)
```

Comment:
I think we have 3 copies of this function now, in `lower_convs_to_matmul.py, subpixel_to_deconv.py` and now here too. I understand it probably got duplicated because the `_` prefix indicates a "private" function, but if these functions are all the same we should consider moving this to one of the utilities modules and making it public to allow its reuse.